lib/block_orthogonalize.cu

```diff
@@ -1,188 +1,211 @@
+      TunableBlock2D(V, chiral_blocks),
+      V(V),
+      B(B),
+      fine_to_coarse(fine_to_coarse),
+      coarse_to_fine(coarse_to_fine),
+      geo_bs(geo_bs),
+      n_block_ortho(n_block_ortho),
+      two_pass(two_pass),
+      iter(0),
+      max(1.0)
+    {
+      if (nColor_ != nColor)
+        errorQuda("Number of colors %d not supported with this precision %lu\n", nColor_, sizeof(bFloat));
 
-        if (V.Location() == QUDA_CUDA_FIELD_LOCATION) {
-#ifdef JITIFY
-          create_jitify_program("kernels/block_orthogonalize.cuh");
-#endif
-        }
-      strcat(aux, compile_type_str(V));
-      strcat(aux, V.AuxString());
       strcat(aux,",block_size=");
 
-      geoBlockSize = 1;
+      aggregate_size = 1;
       char geo_str[16];
       for (int d = 0; d < V.Ndim(); d++) {
-        geoBlockSize *= geo_bs[d];
+        aggregate_size *= geo_bs[d];
         i32toa(geo_str, geo_bs[d]);
         strcat(aux, geo_str);
         if (d < V.Ndim() - 1) strcat(aux, "x");
       }
-      if (geoBlockSize == 1) errorQuda("Invalid MG aggregate size %d", geoBlockSize);
+
+      if (aggregate_size == 1) errorQuda("Invalid MG aggregate size %d", aggregate_size);
+      nBlock = (V.Volume()/aggregate_size) * chiral_blocks;
 
       strcat(aux, ",n_block_ortho=");
       char n_ortho_str[2];
       i32toa(n_ortho_str, n_block_ortho);
       strcat(aux, n_ortho_str);
-
-      if (V.Location() == QUDA_CPU_FIELD_LOCATION) strcat(aux, getOmpThreadStr());
-
-      int chiralBlocks = (nSpin==1) ? 2 : V.Nspin() / spinBlockSize; //always 2 for staggered.
-      nBlock = (V.Volume()/geoBlockSize) * chiralBlocks;
+      strcat(aux, ",mVec=");
+      char mvec_str[3];
+      int active_x_threads = (aggregate_size / 2) * (nSpin == 1 ? 1 : V.SiteSubset());
+      i32toa(mvec_str, tile_size<nColor, nVec>(OrthoAggregates::block_mapper(active_x_threads)));
+      strcat(aux, mvec_str);
+
+      V.Scale(max); // by definition this is true
+      apply(device::get_default_stream());
+      if (two_pass && V.Precision() < QUDA_SINGLE_PRECISION) {  // recompute for more precision
+        iter++;
+        V.Scale(1.05 * max); // the 1.05 gives us some margin
+        apply(device::get_default_stream());
       }
+    }
 
-    /**
-       @brief Helper function for expanding the std::vector into a
-       parameter pack that we can use to instantiate the const arrays
-       in BlockOrthoArg and then call the CPU variant of the block
-       orthogonalization.
-     */
     template <typename Rotator, typename Vector, std::size_t... S>
-    void CPU(const std::vector<ColorSpinorField*> &B, std::index_sequence<S...>) {
-      typedef BlockOrthoArg<Rotator,Vector,nSpin,spinBlockSize,coarseSpin,nVec> Arg;
-      Arg arg(V, fine_to_coarse, coarse_to_fine, QUDA_INVALID_PARITY, geo_bs, n_block_ortho, V, B[S]...);
-      blockOrthoCPU<sumType,RegType,nSpin,spinBlockSize,nColor,coarseSpin,nVec,Arg>(arg);
+    void CPU(const TuneParam &tp, const qudaStream_t &stream,
```

Comment:
not changed in this PR but CPU and GPU and not the greatest function nanmes here. Could probably be fused with a templated function.